subsys/bootloader/cmake/sign.cmake

```diff
@@ -47,13 +47,14 @@ if (CONFIG_SB_PRIVATE_KEY_PROVIDED)
     ${PROJECT_BINARY_DIR}
     USES_TERMINAL
     )
+endif()
 
-  add_custom_target(
+# Public key file target is required for all signing options
+add_custom_target(
     signature_public_key_file_target
     DEPENDS
     ${SIGNATURE_PUBLIC_KEY_FILE}
```

Comment:
You now have a dependency to `SIGNATURE_PUBLIC_KEY_FILE` which is created here:
https://github.com/nrfconnect/sdk-nrf/blob/247e4340bedfb6b2afe45b090bb90911729a0df0/subsys/bootloader/cmake/sign.cmake#L37-L39
but that is only created if `CONFIG_SB_PRIVATE_KEY_PROVIDED` is not an empty string.

Or when a valid file is provided in `CONFIG_SB_SIGNING_PUBLIC_KEY` when `CONFIG_SB_SIGNING_CUSTOM`.
 
If no valid file is given in `CONFIG_SB_PRIVATE_KEY_PROVIDED` or `CONFIG_SB_SIGNING_PUBLIC_KEY`, then this target will always be out of date.

It seems to me more holes exists in https://github.com/nrfconnect/sdk-nrf/blob/master/subsys/bootloader/Kconfig in combination with https://github.com/nrfconnect/sdk-nrf/blob/master/subsys/bootloader/cmake/sign.cmake

Reply:
Perhaps what this is missing is an error check here for verifying that the given public file exists:
https://github.com/nrfconnect/sdk-nrf/blob/247e4340bedfb6b2afe45b090bb90911729a0df0/subsys/bootloader/cmake/sign.cmake#L30-L32

From what I understand from the bootloader Kconfig, `CONFIG_SB_SIGNING_PUBLIC_KEY` is only relevant for custom signing. Since custom signing is intended to only work with public keys, this option must point to an valid public key existing on the system.

`CONFIG_SB_PRIVATE_KEY_PROVIDED` is set by non-custom signing options and enables public key generation via `PUB_GEN_CMD` variables. Those are created in [lines 15-29](https://github.com/nrfconnect/sdk-nrf/blob/247e4340bedfb6b2afe45b090bb90911729a0df0/subsys/bootloader/cmake/sign.cmake#L15-L29) and then called in the referenced custom command: https://github.com/nrfconnect/sdk-nrf/blob/247e4340bedfb6b2afe45b090bb90911729a0df0/subsys/bootloader/cmake/sign.cmake#L37-L43

So as long as we verify the existence of the key in `CONFIG_SB_SIGNING_PUBLIC_KEY`, `SIGNATURE_PUBLIC_KEY_FILE` will be up to date by the time it is used in the script, regardless of signing method.

Reply:
> Perhaps what this is missing is an error check here for verifying that the given public file exists:

Exactly what I was hinting above.

Without a check, user can set `CONFIG_SIGNATURE_PUBLIC_KEY_FILE=missing.pub` and run CMake, and get following output:
```
-- Configuring done
-- Generating done
-- Build files have been written to: /projects/github/ncs/nrf/build
```
but fail when running ninja:
```
$ ninja 
ninja: error: 'modules/nrf/subsys/bootloader/image/missing.pub', needed by 'modules/nrf/subsys/bootloader/image/CMakeFiles/signature_public_key_file_target', missing and no known rule to make it
```

so check for the file during CMake, to ensure an early failure, and then:
1. Fail the CMake stage
2. Print a warning that key file is missing and signing stage is disabled.
    This allows for keeping the build target, and just print warning that public key is missing during signing.

Personally I favor 2., cause that allows for distributing / re-using Kconfig conf files, even without having the public key available.
This allows users to still build the code and test it locally, without having to modify the configuration.

@hakonfam @oyvindronningstad do any of you have preferences for 1. or 2. ?

Also check the behaviour for missing / wrong defined `SIGNATURE_PRIVATE_KEY_FILE`

Reply:
> 2. Print a warning that key file is missing and signing stage is disabled.
>     This allows for keeping the build target, and just print warning that public key is missing during signing.
>
> Personally I favor 2., cause that allows for distributing / re-using Kconfig conf files, even without having the public key available.

That's interesting. I would think an error message would be more suitable for a required option that has an invalid value, but these would break portability?

Reply:
> That's interesting. I would think an error message would be more suitable for a required option that has an invalid value, but these would break portability?

Under normal situations I would completely agree with you, but in terms of signing things may be a little different.
If you look at NCSDK-7277, then you'll see that developers may not always have keys available.
A missing file might be considered an error, I was mixing with the mcuboot case, where the pem file might only contain the public, and not the private part of the key.

Now, this is in fact the public key, so less likely, but private key can easily be something developers don't have and thus an error could be fine.

In general I think we ought to look into a more consistent behavior regarding key handling (and naming).

Today, we actually have:
- `MCUBOOT_SIGNATURE_KEY_FILE`
- `BOOT_SIGNATURE_KEY_FILE`
- `SB_SIGNING_PUBLIC_KEY`
- `SB_SIGNING_KEY_FILE`

As you'll notice in `NCSDK-7277`, the `MCUBOOT_SIGNATURE_KEY_FILE` hold the keypair or just the public key.
If only the public key is available, then signing should be disabled, like seen here: #3580 .

Reply:
> In general I think we ought to look into a more consistent behavior regarding key handling (and naming).

I agree :+1: This is one of the reasons behind the documentation in #3607, which lead me to raising this PR while navigating some of these key-based features.

However, with regards to the scope of this PR: I'm happy to add a warning to catch a missing public signature key file so that the dependency is satisfied.

Reply:
> This is one of the reasons behind the documentation in #3607, which lead me to raising this PR while navigating some of these key-based features.

:+1: 

> I'm happy to add a warning to catch a missing public signature key file so that the dependency is satisfied.

A good start.